sim/core/movement.go

```diff
@@ -0,0 +1,139 @@
+	startTime   time.Duration // starting time of the movement
+	speed       float64       // theoretical movement speed, can be 0
+}
+
+func (action *MovementAction) GetCurrentPosition(sim *Simulation) float64 {
+	return action.srcPosition + float64(sim.CurrentTime-action.startTime)*action.speed/float64(time.Second)
+}
+
+func (unit *Unit) initMovement() {
+	unit.moveAura = unit.GetOrRegisterAura(Aura{
+		Label:     "Movement",
+		ActionID:  ActionID{OtherID: proto.OtherAction_OtherActionMove},
+		Duration:  NeverExpires,
+		MaxStacks: 30,
+
+		OnGain: func(aura *Aura, sim *Simulation) {
+			unit.Moving = true
+		},
+		OnExpire: func(aura *Aura, sim *Simulation) {
+			unit.Moving = false
+		},
+	})
+
+	unit.moveSpell = unit.GetOrRegisterSpell(SpellConfig{
+		ActionID: ActionID{OtherID: proto.OtherAction_OtherActionMove},
+		Flags:    SpellFlagMeleeMetrics,
+
+		ApplyEffects: func(sim *Simulation, target *Unit, spell *Spell) {
+			unit.moveAura.Activate(sim)
+			unit.moveAura.SetStacks(sim, max(int32(unit.DistanceFromTarget), 1))
+		},
+	})
+
+	unit.RegisterResetEffect(func(s *Simulation) {
+		unit.Moving = false
+		unit.movementAction = nil
+	})
+}
+
+func (unit *Unit) MoveTo(moveRange float64, sim *Simulation) {
+	if moveRange == unit.DistanceFromTarget {
+		return
+	}
+
+	moveDistance := moveRange - unit.DistanceFromTarget
+	timeToMove := time.Duration(math.Abs(moveDistance)/unit.GetMovementSpeed()*1000) * time.Millisecond
+	registerMovementAction(unit, sim, unit.GetMovementSpeed()*TernaryFloat64(moveDistance < 0, -1., 1.), sim.CurrentTime+timeToMove)
+}
+
+func (unit *Unit) MoveDuration(duration time.Duration, sim *Simulation) {
+	if duration == 0 {
+		return
+	}
+
+	registerMovementAction(unit, sim, 0., sim.CurrentTime+duration)
+}
+
+func (unit *Unit) UpdatePosition(sim *Simulation) {
+	if !unit.Moving {
+		return
+	}
+
+	unit.DistanceFromTarget = unit.movementAction.GetCurrentPosition(sim)
+	unit.OnMovement(unit.DistanceFromTarget, MovementUpdate)
+
+	yards := max(int32(unit.DistanceFromTarget), 1) // never set to 0 yards as we deactivate the aura
+	if yards != unit.moveAura.GetStacks() {
+		unit.moveAura.SetStacks(sim, yards)
+	}
+}
+
+func (unit *Unit) FinalizeMovement(sim *Simulation) {
+	if !unit.Moving {
+		return
+	}
+
+	unit.UpdatePosition(sim)
+	unit.movementAction = nil
+	unit.moveAura.Deactivate(sim)
+
+	unit.OnMovement(unit.DistanceFromTarget, MovementEnd)
+}
+
+func registerMovementAction(unit *Unit, sim *Simulation, speed float64, endTime time.Duration) {
+	if unit.movementAction != nil {
+		unit.UpdatePosition(sim)
+		unit.movementAction.Cancel(sim)
+		unit.movementAction = nil
+	} else {
+		unit.moveSpell.Cast(sim, unit.CurrentTarget)
+	}
+
+	movementAction := MovementAction{
+		startTime:   sim.CurrentTime,
+		speed:       speed,
+		srcPosition: unit.DistanceFromTarget,
+	}
+
+	movementAction.NextActionAt = endTime
+	movementAction.OnAction = func(sim *Simulation) {
+		unit.FinalizeMovement(sim)
+	}
```

Comment:
So if I'm understanding this correct you only update the distance from target at the start of the movement and at the end of the movement. What happens if I'm at 3 yards and want to move to 8 yards. Will my melee auto attack get disabled when I cross the 5 yards threshold or will it get instantly disabled when i start moving?

Reply:
Ok, i see that you are updating the current position from the APLs callback which also works (maybe even better so every action has the precise current position)

Reply:
Distance is updated before APL actions if the unit is moving to ensure correct positions. Due to the extra conditions in the Attacks if we're in range we attack, if not - we don't. Melee ranged attack will always happen on retail when you're within range limitations.

Given your comment below, I can also disable enable them based on distance during movement udpate.